internal/handler/executor.go

```diff
@@ -134,13 +138,49 @@ func (e *Executor) execHookCommand(w io.Writer) error {
 		// log only envs set by webhook, not global env; otherwise it's leaking secrets to logs
 		"environment", envs,
 		"working_directory", cmd.Dir,
+		"timeout", timeout,
 	)
 	cmd.Stderr = w
 	cmd.Stdout = w
-
+	// handling the timeout
+	if timeout > 0 {
+		// sets the same PGID for the child processes
+		setPGID(cmd)
+		terminationTimer := e.stopProcessWithTimeout(cmd, timeout)
+		err := cmd.Run()
+		// stop timer if command had finished before the timeout reached
+		terminationTimer.Stop()
+		return err
+	}
 	return cmd.Run()
 }
 
+// stopProcessWithTimeout handles termination of the process with configurable timeout
+func (e *Executor) stopProcessWithTimeout(cmd *exec.Cmd, timeout time.Duration) *time.Timer {
+	e.logger.Info("Setting up timeout for current operation", "timeout", timeout)
+
+	return time.AfterFunc(timeout, func() {
+
+		e.logger.Info("Sending SIGTERM because timeout has reached", "timeout", timeout)
+		// attempting to terminate the process with pre-configured timeout
+		if err := sendKillSignal(e, cmd.Process.Pid, syscall.SIGTERM); err != nil {
+			e.logger.Warn("Failed to send SIGTERM, trying SIGKILL instead", "error", err)
+		}
+		// attempting to kill the process with additional timeout on top
+		killingTimer := time.AfterFunc(time.Second*time.Duration(10), func() {
+			if err := sendKillSignal(e, cmd.Process.Pid, syscall.SIGKILL); err != nil {
+				e.logger.Error("Failed to send SIGKILL", "error", err)
+			}
+		})
+		// waiting for the process being exited
+		if processState, err := cmd.Process.Wait(); err != nil && processState != nil {
+			e.logger.Error("Error during process exiting", "error", err)
+		}
+		// stop the timer if process had terminated before timer reached
+		killingTimer.Stop()
```

Comment:
hoist to the timer creation point and refactor like this:

```suggestion
		killingTimer := time.AfterFunc(time.Second*time.Duration(10), func() {
			if err := sendKillSignal(e, cmd.Process.Pid, syscall.SIGKILL); err != nil {
				e.logger.Error("Failed to send SIGKILL", "error", err)
			}
		})
		defer func(t *time.Timer) {
		    // stop the timer if process had terminated before timer reached
  		    if !t.Stop() {
		        <-t.C
		    }
		}(killingTimer)

		// waiting for the process being exited
		if processState, err := cmd.Process.Wait(); err != nil && processState != nil {
			e.logger.Error("Error during process exiting", "error", err)
		}
```